man/delrec.3
.TH DELREC 3 "12 September 1995"
.FB
.SH NAME
delrec, b_delrec \- delete a record
.SH SYNOPSIS
#include <firstbase/fb.h>
.sp 1
fb_delrec(db)
.br
fb_database *db;
.sp 1
fb_b_delrec(db)
.br
fb_database *db;
.br
.SH DESCRIPTION
As explained in the \fIdatabase\fP(4)
manual page, a database structure contains
all of the elements needed for an entire database, including an array of
field structures, all file names and descriptors, and index information.
Altogether, the array of field structures can hold one complete record.
.PP
These records are retrieved from the database file and stored into the
field structures using \fIgetrec\fP(3) or \fIgetxrec\fP(3).
Once a record has been
loaded into these field structures, \fIfb_delrec\fP is used
to mark current \fIdb\fP record as deleted.
.PP
\fIFb_delrec\fP will perform the following sequence of events:
.sp 1
.PD 0
.TP 3
1)
The database header variables \fIdb->reccnt\fP and \fIdb->delcnt\fP are read.
.TP 3
2)
An asterisk is stored in the last field slot of the array
of field structures.
This last field is always set up by the \fIFirstBase\fP system and never needs to be
defined within a database dictionary.
True database fields, as defined in a database dictionary,
are indexed within the array
of field structures beginning at 0 and going through the number of fields
minus 1. The last, undefined field structure is used solely for marking
a record as deleted.
Most \fIFirstBase\fP tools and subroutines will ignore records marked for deletion.
.TP 3
3)
The deleted
record count is bumped to \fIdb->delcnt\fP + 1 and the record is replaced in
the database using the \fIputrec\fP(3) routine.
.TP 3
4)
The database header information is written back into the database.
.PD
.PP
Note that \fIfb_delrec\fP does not attempt to modify any
normal (flat) indexes, but btree indexes defined as autoindexes are maintained.
See \fIindex\fP(4) and \fIautoindex\fP(4) for details.
.PP
Sometimes it is desirable to delete many records as fast as possible.
Since \fIfb_delrec\fP will lock the database,
synchronize the data files, and unlock the database
for each record, it can be inefficient. The bulk record processing
mechanism can be used in these cases. (See \fIbulkrec\fP(3)).
.PP
The bulk processing of records consists of three steps. First,
\fIfb_bulkrec_begin\fP is called before any processing. This preliminary
call will lock and synchronize the database, and read the database header.
Then, \fIfb_b_delrec\fP and/or \fIfb_b_addrec\fP are called repeatedly 
as needed.
Finally, \fIfb_bulkrec_end\fP is used to write the header,
synchronize the data files, and unlock the database.
.SH CLIENT/SERVER
The \fIfb_delrec\fP and \fIfb_b_delrec\fP routines will transparently
make calls to \fIfb_delrec_clnt\fP and \fIfb_b_delrec_clnt\fP
when the global \fIFirstBase\fP variable \fIcdb_use_rpc\fP is set to one.
See \fIfb_clnt_create\fP(3) for more details.
.SH SEE ALSO
getrec(3), getxrec(3), opendb(3), addrec(3), bulkrec(3),
database(4), dbpack(1), dbundo(1).
.SH DIAGNOSTICS
\fIFb_delrec\fP will return the error code FB_ERROR if any of its reads,
writes, or seeks fail. An FB_AOK signal is returned on correct completion.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/librec/delrec.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: delrec.c,v 9.1 2001/02/16 19:34:19 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Delrec_sid[] = "@(#) $Id: delrec.c,v 9.1 2001/02/16 19:34:19 john Exp $";
#endif

#include <fb.h>
#include <fb_ext.h>

extern short int cdb_use_rpc;
extern short int cdb_returnerror;
extern short int cdb_error;
extern short int cdb_secure;

/*
 * delrec - delete the loaded record from the *dp.
 *	- decrement the header and restore.
 *	- NOT responsibility of delrec to do mutual exclusion anywhere.
 */

   fb_delrec(dp)
      fb_database *dp;
      
      {
         int st = FB_AOK;

#if RPC
         if (cdb_use_rpc)
            return(fb_delrec_clnt(dp));
#endif
         if (dp->rec < 1 || dp->rec > dp->reccnt)
            return(FB_ERROR);
	 fb_lock_head(dp);
	 fb_setdirty(dp, 1);
	 fb_allsync(dp);
	 st = fb_gethead(dp);
	 if (st == FB_ERROR){
            cdb_error = FB_READ_ERROR;
            if (cdb_returnerror){
	       fb_lerror(FB_READ_ERROR, SYSMSG[S_BAD_HEADER], dp->dbase);
               return(FB_ERROR);
               }
	    fb_xerror(FB_READ_ERROR, SYSMSG[S_BAD_HEADER], dp->dbase);
            }
         if (cdb_secure)
            fb_recmode(dp, CHAR_STAR, -1, -1, NIL);	/* update del only */
         else{
            /* make sure there is enogh room to store the '*' */
            fb_store(dp->kp[dp->nfields], SYSMSG[S_STRING_BLANK], dp);
            dp->kp[dp->nfields]->fld[0] = CHAR_STAR;
            }
	 if (fb_putrec(dp->rec, dp) == FB_ERROR){
            cdb_error = FB_WRITE_ERROR;
	    fb_lerror(FB_WRITE_ERROR, SYSMSG[S_BAD_DATA], dp->dbase);
            return(FB_ERROR);
            }
	 if (fb_puthead(dp->fd, dp->reccnt, ++dp->delcnt) == FB_ERROR){
            cdb_error = FB_WRITE_ERROR;
	    fb_lerror(FB_WRITE_ERROR, SYSMSG[S_BAD_HEADER], dp->dbase);
            return(FB_ERROR);
            }
	 fb_setdirty(dp, 0);
         if (dp->b_tree == 1)
            st = fb_delidx(dp, dp->rec);
	 fb_allsync(dp);
	 fb_unlock_head(dp);
	 return(st);
      }
